File: code/cpp/babilonska.cpp
// Metoda babilonska (metoda Herona) obliczania pierwiastka kwadratowego.
// Kompilacja: g++ -O2 -std=c++17 babilonska.cpp -o babilonska
#include <iostream>
#include <cmath>
#include <iomanip>

// Oblicza sqrt(S) metoda babilonska z zadana dokladnoscia (domyslnie 1e-15).
double pierwiastekBabilonski(double S, double eps = 1e-15) {
    if (S < 0) return NAN;
    if (S == 0) return 0.0;

    double x = S;                 // przyblizenie poczatkowe
    double poprzednie;

    do {
        poprzednie = x;
        x = 0.5 * (x + S / x);    // wzor iteracyjny Herona
    } while (std::fabs(x - poprzednie) > eps);

    return x;
}

int main() {
    double liczby[] = {2.0, 10.0, 144.0, 1000.0, 999999.0};

    std::cout << std::fixed << std::setprecision(15);
    for (double S : liczby) {
        double wynik = pierwiastekBabilonski(S);
        double wzorcowy = std::sqrt(S);
        std::cout << "sqrt(" << S << ")\n"
                  << "  metoda babilonska : " << wynik << "\n"
                  << "  std::sqrt         : " << wzorcowy << "\n"
                  << "  roznica           : " << std::fabs(wynik - wzorcowy) << "\n\n";
    }
    return 0;
}

File: docs/code/python/babilonska.py
"""Metoda babilonska (metoda Herona) obliczania pierwiastka kwadratowego."""
import math


def pierwiastek_babilonski(s: float, eps: float = 1e-15) -> float:
    """Oblicza sqrt(s) metoda babilonska z zadana dokladnoscia."""
    if s < 0:
        raise ValueError("Liczba pod pierwiastkiem nie moze byc ujemna")
    if s == 0:
        return 0.0

    x = s
    while True:
        poprzednie = x
        x = 0.5 * (x + s / x)
        if abs(x - poprzednie) <= eps:
            return x


if __name__ == "__main__":
    for liczba in (2, 10, 144, 1000, 999999):
        wynik = pierwiastek_babilonski(liczba)
        wzorcowy = math.sqrt(liczba)
        print(f"sqrt({liczba})")
        print(f"  metoda babilonska : {wynik:.15f}")
        print(f"  math.sqrt         : {wzorcowy:.15f}")
        print(f"  roznica           : {abs(wynik - wzorcowy):.15f}\n")
